model/job-management/src/main/proto/beam_artifact_api.proto

```diff
@@ -31,8 +31,77 @@ option java_outer_classname = "ArtifactApi";
 
 import "beam_runner_api.proto";
 
-// A service to stage artifacts for use in a Job.
+// A service to retrieve artifacts for use in a Job.
+service ArtifactRetrievalService {
+  // Resolves the given artifact reference into one or more simpler artifact
+  // references (e.g. a Maven dependency into a (transitive) set of jars.
+  // If no further simplification is possible, returns the original artifacts,
+  // at which point, all artifacts must be gettable.
+  rpc ResolveArtifact(ResolveArtifactRequest) returns (ResolveArtifactResponse);
+
+  // Retrieves the given artifact as a stream of bytes.
+  rpc GetArtifact(GetArtifactRequest) returns (stream GetArtifactResponse);
+}
+
+// A service that allows the client to act as an ArtifactRetrievalService,
+// for a particular job with the server initiating requests and receiving
+// responses.
+// A client calls the service with an ArtifactResponseWrapper that has the
+// staging token set, and thereafter responds to the server's requests.
 service ArtifactStagingService {
+  rpc ReverseArtifactRetrievalService(stream ArtifactResponseWrapper)
```

Comment:
The style of this stream won't allow for parallel *requests* to be send to the client since there is no *id* mechanism to diffentiate the responses since every request will have to have wait for the *is_last* to be set before the next *request* is sent out. If you want to support parallel *requests*, you also need to have a place to provide errors outside of the gRPC status that can be returned since you'll want to identify which *request* had a problem.

Also, if there are any errors for a single *request*, the whole RPC needs to be terminated since there is no way to say which *request* failed.

Since Java has typically a lot of jars, it is likely that we'll want to be able to parallelize this.

Reply:
Good question. As discussed offline, multiple requests can be pushed before waiting for a response, so we should be able to fill the network buffer. Either the server or client can choose to parallelize on its backend store as needed. 

I added an error field to the resolve_artifacts field as a runner may attempt to opportunistically consolidate environments, and then fall back to doing each separately.